### ISSUE TITLE
Add option to relinquish user-based publish privileges for projects with Trusted Publishing enabled

### ISSUE THREAD

**Issue by sethmlarson** [pypi/warehouse#14300]
**What's the problem this feature will solve?**

Currently being an admin/maintainer of a PyPI project forces your account to have the "publish" privilege. If Trusted Publishers is enabled, having user publish privileges for a project is only additional attack surface area for malware to be published. Opting-in to "Trusted Publisher" publishing only reduces that attack surface to being able to thwart GitHub Environments or access to GitHub accounts.

**Describe the solution you'd like**

If a project has trusted publishing enabled and has successfully published once with the trusted publisher (ie the trusted publisher isn't pending) then add an additional hardening option in the project settings to relinquish user publishing privileges for a project.

With user publishing privileges relinquished:
- Admins/maintainers of the project can't publish using username/password or API tokens
- New trusted publisher configuration can't be added. Existing trusted publisher configuration can't be removed completely, must always have at least 1 trusted publisher configured.
- Admins/maintainers can still yank releases as normal.

**Additional context**

- This will add additional support burden to PyPI if users mis-configure their publishing or at a later time want to migrate to a new service (ie GitHub -> GitLab)

**Comment by di** [comment]
Would this effectively make Trusted Publishing permanent for the project name? Would there ever be a way for the project to disable Trusted Publishing without admin intervention?

(Probably worth considering this in the context of https://github.com/pypi/warehouse/issues/13366 as well).

**Comment by sethmlarson** [comment]
That would be the intention yes, once this option is enabled the name would be permanent until a support request to disable it went through. I can also see a future where one could "upgrade" their Trusted Publisher configuration into a same owner/name/workflow configured "Trusted Builder" once we start looking at SLSA.

**Comment by dstufft** [comment]
I'm not sure I understand what this gains? Even if someone disables user level publishing, they can still add new trusted publishers that can release instead, and it's not like ``dstufft/foo`` vs ``sethmlarson/foo`` on github is a meaningful security boundary.

**Comment by woodruffw** [comment]
> New trusted publisher configuration can't be added. Existing trusted publisher configuration can't be removed completely, must always have at least 1 trusted publisher configured.

This in particular might cause a lot of support load: I could see users wanting to slightly tweak their trusted publishers (e.g. `release.yml -> publish.yml`, changing a GitHub username to an owning org, etc.).

Maybe a (slightly) weaker version of this makes sense? Something like:

* A project can be put into "publisher only" mode, which is permanent (modulo admin intervention, e.g. for takeover);
* In publisher only mode, releases can only be made through trusted publishers; any user- or project-scoped tokens that previously covered the project cease to be valid for it;
* The project must always have at least one trusted publisher configured; project owners can modify the trusted publisher at will.

So basically the same as the original proposal, except that the original trusted publisher configuration can be replaced 🙂 

That still leaves account takeovers/new publishers as an attack vector however, per @dstufft's point.

**Comment by davidism** [comment]
There's also a problem if a project wants to migrate off GitHub in the future and trusted publishing isn't available at the new location.

**Comment by davidism** [comment]
I think it would still serve a useful purpose to be able to toggle user/token publishing from the management page. Disabling it would prevent accidental publishing. Enabling it could be useful in unusual circumstances, and should send an email notification like other changes do.

**Comment by dstufft** [comment]
Currently API tokens are only good for publishing, if you don't want to accidentally publish you could presumably just not create any API tokens? I guess if you want to create new projects from the command line, without enabling uploading projects?

I guess I'm not like, inherently against the idea, I'm just having a hard time understanding what we actually gain.

**Comment by davidism** [comment]
I have a community organization account, similar to Jazzband if you're familiar with it, where we collect popular Flask extensions for shared community maintenance. Some users have "maintainer" status so they can make releases instead of only commit access. Many of the projects aren't yet set up for trusted publishing either, so as they're added we can make sure maintainers don't still manually publish. It would be nice for them to be able to get access to PyPI and manual publishing in case of some sort of issue (such as me not being around for some reason), but for normal circumstances keep it to only trusted publishing.

**Comment by dstufft** [comment]
Ah okay. That makes a bit more sense then, I was still thinking in terms of security boundary and whole accounts at a time rather than thinking of it as a transitional or more narrow scope thing.

I don't personally have any opposition to a flag on a project to make it trusted publishing only, but it should be something that can be toggled on and off rather than some one way switch.

**Comment by ristomcgehee** [comment]
Having the ability to toggle on/off a "trusted publishing only" mode sounds like the right approach to me. If that's the consensus, I'm willing to start work on this issue.

**Comment by mrgrain** [comment]
> Having the ability to toggle on/off a "trusted publishing only" mode sounds like the right approach to me. If that's the consensus, I'm willing to start work on this issue.

I think this is the right approach. The package itself should be able to decide in which ways it can be published. A toggable "trusted publishing only" mode would certainly be the easiest way of achieving this. 

Without this, my concern is that an account level API token might be compromised. And because PyPI allows to publish the package using a token, the whole Trusted Publishing setup is bypassed.

On npmjs.com there is this setting to achieve a similar result:

<img width="704" height="203" alt="Image" src="https://github.com/user-attachments/assets/2c958c8f-06c4-46fb-a226-fd5e8327a48a" />

(Publishing by 2FA-auth'd users makes sense, because with that level of access, an actor could also go and just remove trusted publishing config)

---

Alternatively an option to exclude packages from API tokens would work as well (the inverse of the current "scope") option. But that's less preferable to us.

**Comment by woodruffw** [comment]
I think the recent LiteLLM and telnyx compromises have made this relevant/salient again: neither was using TP but *if* they were, a lingering long-lived API credential could still have caused significant issues (despite the upstream believing that they're safe).

The Rust ecosystem's implementation of Trusted Publishing does something pretty similar to what's proposed above: users can _optionally_ disable API token publishing, at which point any previously issued long-lived API tokens become ineffective, even if not formally destroyed/revoked.

On top of that, it might be nice to have a cooldown period for de-toggling that protection, similar to what's being proposed in #19770 for account recovery codes: if a project has "TP only" enabled and an project owner disables it, the disablement should (1) notify all other owners, and (2) take some period of time to take effect (24/48/72h?).

**Comment by Jackenmen** [comment]
Personally, my issue is also that any maintainer can make a new token scope for the entire account (which, to my knowledge, includes all orgs that the account is part of) *but* it's not going to show up in the organisation/project's security history (and therefore won't email other people in the organisation).

If I could disable user tokens for a project (or an entire org), then, *if* someone wanted to use an entire account token, they would at least have to reenable it for the project (which could also have an additional delay as suggested by William in the comment above), creating an event in the security history for it.

**Comment by woodruffw** [comment]
I've been thinking about this more recently, and I think we should do it: other package registries (notably crates.io for Rust) have adopted a similar pattern with success. For example, here's a random package on crates.io that I maintain that shows the control:

<img width="973" height="318" alt="Image" src="https://github.com/user-attachments/assets/45c990cd-2dcc-441f-9cea-51a40331bdb4" />

A very rough idea:

1. The default remains the same;
2. Users can opt into requiring TP only for projects that have one or more publishers configured;
3. Once that requirement is enabled, users can only disable it after a 72h (96h?) cooldown. This will help with the admittedly niche scenario in which an attacker manages full ATO on PyPI, including MFA, and tries to force publishing by inserting their own API credentials.
